components/higher-order/with-state/README.md

```diff
@@ -0,0 +1,32 @@
+withState
+=========
+
+`withState` is a React [higher-order component](https://facebook.github.io/react/docs/higher-order-components.html) which enables a function component to have internal state.
+
+Wrapping a component with `withState` provides state as props to the wrapped component, along with a `setState` updater function.
+
+## Usage
+
+```jsx
+/**
+ * WordPress dependencies
+ */
+import { withState } from '@wordpress/components';
+
+function MyCounter( { count, setState } ) {
+	return (
+		<>
+			Count: { count }
+			<button onClick={ () => setState( { count: count + 1 } ) }>
```

Comment:
Just noting that React docs discourage using `setState( { count: count + 1 } )`:

> React may batch multiple `setState()` calls into a single update for performance.
>
> Because `this.props` and `this.state` may be updated asynchronously, you should not rely on their values for calculating the next state.

They suggest using the following:
```
setState( count => count + 1 )
```

See https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous. I think we should follow those guidelines in here, and in the unit test.

Reply:
>I think we should follow those guidelines in here, and in the unit test.

Updated in rebased 6fba36df84cb1bf0469b1e1effec8824752a0d21.